TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+{
+    [self pushPathComponent:key];
+    NSArray *array = schema[key];
+    [self checkObjectIsArrayWithAtLeastOneItem:array allowNil:YES];
+    [self popPathComponent];
+    return array;
+}
+
+
+// Keyword: required
+// Example: { "required" : [ "name" ] }
+- (NSArray *)parseArrayOfUnqiueStringsFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSArray *array = schema[key];
+    [self checkObjectIsArrayWithAtLeastOneItem:array allowNil:YES];
+
+    NSMutableSet *uniqueStrings = [[NSMutableSet alloc] init];
+
+    [array enumerateObjectsUsingBlock:^(NSString *string, NSUInteger index, BOOL *stop) {
+        [self pushPathComponent:@(index)];
+        [self checkObject:string isOfClass:[NSString class] allowNil:NO];
+        if ([uniqueStrings containsObject:string]) {
+            NSString *reason = [NSString stringWithFormat:@"Invalid value for key %@. Expected array of unique strings.", key];
+            [self failWithErrorCode:TWTJSONSchemaParserErrorCodeRequiresUniqueness reason:reason object:string];
+        }
+
+        [uniqueStrings addObject:string];
+        [self popPathComponent];
+    }];
+
+    [self popPathComponent];
+    return array;
+}
+
+
+// Keyword: items, anyOf, allOf, oneOf
+// Example: { "items" : [ {}, {} ] }
+/*
+ { "anyOf": [
+ { "type": "string", "maxLength": 5 },
+ { "type": "number", "minimum": 0 } ]
+ }
+ */
+- (NSArray *)parseArrayOfSchemasFromSchema:(NSDictionary *)schema forKey:(NSString *)key
+{
+    [self pushPathComponent:key];
+    NSArray *array = schema[key];
+    [self checkObjectIsArrayWithAtLeastOneItem:array allowNil:YES];
+
+    NSMutableArray *schemaNodes = [[NSMutableArray alloc] init];
+    [array enumerateObjectsUsingBlock:^(NSDictionary *itemSchema , NSUInteger index, BOOL *stop) {
+        [self pushPathComponent:@(index)];
+        [schemaNodes addObject:[self parseSchema:itemSchema]];
+        [self popPathComponent];
+    }];
+
+    [self popPathComponent];
+    return schemaNodes;
+}
+
+
+// Keyword: properties, patternProperties
+// Example: { "properties" : { "hello" : {}, "world" : {} } }
+//          { "patternProperties" : { "^h" : {}, "r" : {} } }
+- (NSArray *)parseDictionaryOfSchemasFromSchema:(NSDictionary *)schema forKey:(NSString *)key nodeClass:(Class)className
+{
+    NSParameterAssert([className isSubclassOfClass:[TWTJSONSchemaKeyValuePairASTNode class]]);
+
+    [self pushPathComponent:key];
+    NSDictionary *nestedSchema = schema[key];
+    [self checkObject:nestedSchema isOfClass:[NSDictionary class] allowNil:YES];
+
+    NSMutableArray *propertyNodes = [[NSMutableArray alloc] init];
+    [nestedSchema enumerateKeysAndObjectsUsingBlock:^(id propertyKey, id object, BOOL *stop) {
+        [self pushPathComponent:propertyKey];
+
+        [self checkObject:propertyKey isOfClass:[NSString class] allowNil:NO]; // Defined by JSON syntax, not schema definition
```

Comment:
This check is unnecessary then, right? You’re assuming you only get JSON schemas as input?